### PR TITLE
[RST-1645] documentation #3

### PR DESCRIPTION
**Pull request by svwilliams**
Documentation for the Constraints class

https://github.com/locusrobotics/fuse/blob/RST-1645-documentation-3/doc/Constraints.md

### CODE REVIEW

**Review by ayrton04 — APPROVED** [review]
Really nice.

File: doc/Constraints.md [hunk 1]
```diff
@@ -0,0 +1,414 @@
+# Constraints
+
+The concept of a Constraints has many names: constraints, cost functions, factors, probably many others. At the most
```
Comment (matthewbwhitaker) [inline doc/Constraints.md]: "concept of Contraints" or "concept of a Constraint"

File: doc/Constraints.md [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+basic level, a Constraint accepts one or more Variable values and produces a score. The Optimizer will then search
+for the values of all the Variables that minimizes the total score. This search is performed by a nonlinear
+least-squares solver. The fuse stack employees [Google's Ceres Solver](http://ceres-solver.org/) as that least-squares
```
Comment (matthewbwhitaker) [inline doc/Constraints.md]: "employs"

File: doc/Constraints.md [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+solver. As a consequence, the fuse_core::Constraint plugin classes are basically just vehicles for creating the
+required Ceres Solver objects.
+
+## Constraint Modeling
+
+Ceres Solver provides an excellent [tutorial](http://ceres-solver.org/nnls_modeling.html#introduction) on system
+modeling for least-squares optimization. The following are some important highlights, but reading the Ceres Solver
+documentation it is highly recommended.
```
Comment (matthewbwhitaker) [inline doc/Constraints.md]: ~it~

File: doc/Constraints.md [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+
+The nonlinear least-squares system used be the Ceres Solver is written as:
```
Comment (matthewbwhitaker) [inline doc/Constraints.md]: "used by"

File: doc/Constraints.md [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+
+![least squares equation](constraints-least_squares_equation.gif)
+
+In Ceres Solver parlance, &rho;() is called a "loss function". f() is called a "cost function", which accepts one or
+more inputs, x. And the inputs, x, are called "parameter blocks". The "parameter blocks" themselves may be
+single-dimensional or multi-dimensional vectors. The least-squares system consists of many &rho;(f()<sup>2</sup>)
+terms, and the solver's job is to find the specific values of the all the inputs that minimize the sum of the squared
```
Comment (matthewbwhitaker) [inline doc/Constraints.md]: "values of ~the~ all the"

File: doc/Constraints.md [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+In fuse, a Constraint class models one of the &rho;(f()<sup>2</sup>) terms.
+
+### Parameter Blocks
+
+The concept of the "parameter block" is modeled by the [Variable](Variables.md) classes in fuse. The Constraint merely
+needs to define which specific Variable identities are to be used as the input to modeled cost function. This is done
+by accepting, at a minimum, the UUID of each involved Variable.
+
+The derived Constraints provided by the [fuse_constraints](../fuse_constraints) package go one step further, and
+require instances of full Variable types rather than just the Variable's UUID. This is done solely enforce type-safety;
```
Comment (matthewbwhitaker) [inline doc/Constraints.md]: "solely to enforce"

File: doc/Constraints.md [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+
+### Cost Function
+
+The "cost function" is the main component of the Constraint object. It is responsible for computing the cost to be
+minimized by the Ceres Solver optimizer. The cost function must implement some sort of equation to generate a score
+for arbitrary input values. In its most generic form, that equation is written simply as:
+
+![generic cost equation](constraints-generic_cost_equation.gif)
+
+where f() is th ecost function, x<sub>1</sub> through x<sub>n</sub> are the input Variables, each of which may contain multi-dimensional data, and r<sub>i</sub> are one or more dimensions of the computed costs. In Ceres Solver
```
Comment (matthewbwhitaker) [inline doc/Constraints.md]: "the cost"

File: doc/Constraints.md [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+practice, loss functions are used to reduce the impact of incorrect or outlier constraints that are included in the
+least-squares problem. By default, the loss function is the identity function, which means that deviations from the
+ideal input values for that cost function will have a quadratic effect on the optimization. If erroneous constraints
+may be added to the system, this quadratic effect can lead to poor optimized results. Various loss functions are
+provided by Ceres Solver that reduce the impact of cost functions with very high costs. The Huber Loss Function is a
+common choice, which applies a quadratic effect for small costs, and a linear effect for large costs. And of coarse,
+you can implement your own loss functions if the desired function is not available from Ceres Solver itself.
+
+![Loss Function Effects](http://ceres-solver.org/_images/loss.png)
+Example Loss Functions available in Ceres Solver
```
Comment (matthewbwhitaker) [inline doc/Constraints.md]: Don't forget you could use HTML here to center the image and caption with `<center>`.

File: doc/Constraints.md [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+  All derived classes must implement the `costFunction()` method, and return a valid Ceres Solver `CostFunction`
+  pointer. Ceres Solver will take ownership of the loss function pointer and delete it when appropriate. There are
+  a few `CostFunction` objects available within the Ceres Solver code base, and several additional ones available in
+  the [fuse_constraints](../fuse_constraints) package. However, it is expected that most derived classes will need to
+  implement their `ceres::CostFunction` as well. This is how the author of the derived class is able to model the
+  specific behavior of a sensor or vehicle. See the
+  [Ceres documentation](http://ceres-solver.org/nnls_modeling.html#costfunction) for details on implementing a
+  `ceres::CostFunction`. Additionally, see the [documentation](http://ceres-solver.org/derivatives.html) about
+  implementing derivatives within Ceres Solver. Ceres Solver supports a powerful automatic derivative system, in
+  addition to numerical and analytic derivatives. This is one of the features drove the selection of Ceres Solver as
```
Comment (matthewbwhitaker) [inline doc/Constraints.md]: "that drove"

File: doc/Constraints.md [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+measure the full robot pose in a global coordinate frame. Things like GPS or scan matching to a prior map could be
+modeled this way. Our cost function will follow the "observation model", where the input variables will be used to
+predict the sensor measurement, and the cost will be the different between the measured and the prediction normalized
+by the measurement uncertainty.
+
+![2D pose prior cost equation](constraints-pose_2d_prior_equation.gif)
+
+We will make use of Ceres Solver's automatic derivative system to compute the Jacobians. For that to work, we must
+implement the cost function equation as a functor object (has an `operator()` method). To compute the cost, our
+functor we need access to the measurement vector (3x1 vector) and the measurement uncertainty. Holding the uncertainty
```
Comment (matthewbwhitaker) [inline doc/Constraints.md]: "functor will(?) need access"

File: doc/Constraints.md [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+  fuse_core::Constraint::UniquePtr clone() const override
+  {
+    return AbsolutePose2DStampedConstraint::make_unique(*this);
+  }
+```
+
+And finally, the required `costFunction()` implementation. We use a combination of the `ceres::AutoDiffCostFunction`
+class and our custom cost functor to create the output `ceres::CostFunction` pointer. The full details of the
+`ceres::AutoDiffCostFunction` are available [here](http://ceres-solver.org/nnls_modeling.html#autodiffcostfunction).
+But briefly, the `ceres::AutoDiffCostFunction` template parameters are: the functor class used to compute the cost,
```
Comment (matthewbwhitaker) [inline doc/Constraints.md]: Either listify, or separate elements with ";".

**Review by matthewbwhitaker — APPROVED** [review]
Excellent docs.

File: doc/Constraints.md (hunk 8 above)
Comment (svwilliams) [reply doc/Constraints.md]: You say "forget" like I knew I could do that.

**Pull request merged**